File: bot/src/walrus.ts
import { CONFIG } from "./config.js";

interface FormDefinition {
  id: string;
  title: string;
  description: string;
  owner: string;
  createdAt: string;
  fields: unknown[];
  settings: unknown;
  tweetUrl?: string;
  tweetAuthor?: string;
}

interface WalrusResponse {
  newlyCreated?: {
    blobObject: {
      blobId: string;
    };
  };
  alreadyCertified?: {
    blobId: string;
  };
}

/**
 * Store a form definition on Walrus and return the blob ID.
 */
export async function storeFormOnWalrus(
  form: FormDefinition
): Promise<string> {
  const body = JSON.stringify(form);

  const response = await fetch(
    `${CONFIG.walrus.publisherUrl}/v1/blobs?epochs=5`,
    {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body,
    }
  );

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`Walrus upload failed: ${response.status} ${text}`);
  }

  const data = (await response.json()) as WalrusResponse;

  const blobId =
    data.newlyCreated?.blobObject.blobId || data.alreadyCertified?.blobId;

  if (!blobId) {
    throw new Error("No blobId in Walrus response");
  }

  return blobId;
}

/**
 * Get the public URL for a Walrus blob.
 */
export function getBlobUrl(blobId: string): string {
  return `${CONFIG.walrus.aggregatorUrl}/v1/blobs/${blobId}`;
}

File: enclave/walrus.rs
use reqwest::Client;
use tracing::info;

use super::types::{FormDefinition, WalrusResponse};

pub async fn store_form_on_walrus(
    publisher_url: &str,
    form: &FormDefinition,
) -> Result<String, String> {
    let client = Client::new();
    let url = format!("{}/v1/blobs?epochs=5", publisher_url);

    let json_body =
        serde_json::to_string(form).map_err(|e| format!("Failed to serialize form: {e}"))?;

    info!("Uploading form '{}' to Walrus...", form.title);

    let resp = client
        .put(&url)
        .header("Content-Type", "application/json")
        .body(json_body)
        .send()
        .await
        .map_err(|e| format!("Walrus upload failed: {e}"))?;

    let status = resp.status();
    let body = resp
        .text()
        .await
        .map_err(|e| format!("Failed to read Walrus response: {e}"))?;

    if !status.is_success() {
        return Err(format!("Walrus upload error {}: {}", status, body));
    }

    let walrus_resp: WalrusResponse =
        serde_json::from_str(&body).map_err(|e| format!("Failed to parse Walrus response: {e}"))?;

    let blob_id = if let Some(newly_created) = walrus_resp.newly_created {
        newly_created.blob_object.blob_id
    } else if let Some(already_certified) = walrus_resp.already_certified {
        already_certified.blob_id
    } else {
        return Err("No blob ID in Walrus response".to_string());
    };

    info!("Form uploaded to Walrus: {}", blob_id);
    Ok(blob_id)
}
